File: ios/BatteryModule.m
//
//  BatteryModule.m
//  financialempire
//
//  Created by Kamil Szerląg on 10/09/2025.
//

#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

@interface RCT_EXTERN_MODULE(BatteryModule, RCTEventEmitter)


// RCT_EXTERN_MODULE - is crucial for React Native to properly register the module.
RCT_EXTERN_METHOD(getBatteryLevel:(RCTPromiseResolveBlock)resolve rejector:(RCTPromiseRejectBlock)reject)

@end

File: ios/BatteryModule.swift
//
//  BatteryModule.swift
//  financialempire
//
//  Created by Kamil Szerląg on 10/09/2025.
//

import Foundation
import UIKit
import React

@objc(BatteryModule)
class BatteryModule: RCTEventEmitter {
  
  override init() {
    super.init()
    UIDevice.current.isBatteryMonitoringEnabled = true
    NotificationCenter.default.addObserver(self, selector: #selector(self.batteryLevelDidChangeNotification), name: UIDevice.batteryLevelDidChangeNotification, object: nil
    )
  }

  deinit {
    NotificationCenter.default.removeObserver(self)
  }
  
  // handle events
  @objc func batteryLevelDidChange(notification: NSNotification) {
    let level = UIDevice.current.batteryLevel
    if level >= 0 {
      sendEvent(withName: "batteryLevelDidChange", body: Int(level * 100))
    }
  }
  
  // RCTEventEmitter
  override func supportedEvents() -> [String]! {
    return ["batteryLevelDidChange"]
  }
  
  @objc
  func getBatteryLevel(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    let level = UIDevice.current.batteryLevel
    if level < 0 {
      reject("E_BATTERY", "Battery level not available", nil)
    } else {
      resolve(Int(level * 100))
    }
  }

  @objc
  static func requireMainQueue() -> Bool {
    return true;
  }

}
